core/cmd/dkg_setup.go

```diff
@@ -0,0 +1,245 @@
+package cmd
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"log"
+	"os"
+	"time"
+
+	"github.com/pelletier/go-toml"
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/chainlink/core/services/job"
+	"github.com/smartcontractkit/chainlink/core/services/pg"
+	"github.com/smartcontractkit/chainlink/core/shutdown"
+	"github.com/smartcontractkit/chainlink/core/static"
+	clipkg "github.com/urfave/cli"
+)
+
+type RunNodeDKGPayload struct {
+	OnChainPublicKey  string
+	OffChainPublicKey string
+	ConfigPublicKey   string
+	PeerID            string
+	Transmitter       string
+	DkgEncrypt        string
+	DkgSign           string
+}
+
+func (cli *Client) RunNodeDKG(c *clipkg.Context) (*RunNodeDKGPayload, error) {
+	lggr := cli.Logger.Named("RunNode")
+
+	err := cli.Config.Validate()
+	if err != nil {
+		return nil, errors.Wrap(err, "config validation failed")
+	}
+
+	lggr.Infow(fmt.Sprintf("Configuring Chainlink Node FOR DKG %s at commit %s", static.Version, static.Sha), "Version", static.Version, "SHA", static.Sha)
+
+	ldb := pg.NewLockedDB(cli.Config, lggr)
+
+	// rootCtx will be cancelled when SIGINT|SIGTERM is received
+	rootCtx, cancelRootCtx := context.WithCancel(context.Background())
+
+	// cleanExit is used to skip "fail fast" routine
+	cleanExit := make(chan struct{})
+	var shutdownStartTime time.Time
+	defer func() {
+		close(cleanExit)
+		if !shutdownStartTime.IsZero() {
+			log.Printf("Graceful shutdown time: %s", time.Since(shutdownStartTime))
+		}
+	}()
+
+	go shutdown.HandleShutdown(func(sig string) {
+		lggr.Infof("Shutting down due to %s signal received...", sig)
+
+		shutdownStartTime = time.Now()
+		cancelRootCtx()
+
+		select {
+		case <-cleanExit:
+			return
+		case <-time.After(cli.Config.ShutdownGracePeriod()):
+		}
+
+		lggr.Criticalf("Shutdown grace period of %v exceeded, closing DB and exiting...", cli.Config.ShutdownGracePeriod())
+		// LockedDB.Close() will release DB locks and close DB connection
+		// Executing this explicitly because defers are not executed in case of os.Exit()
+		if err = ldb.Close(); err != nil {
+			lggr.Criticalf("Failed to close LockedDB: %v", err)
+		}
+		if err = cli.CloseLogger(); err != nil {
+			log.Printf("Failed to close Logger: %v", err)
+		}
+
+		os.Exit(-1)
+	})
+
+	if err = ldb.Open(rootCtx); err != nil {
+		return nil, cli.errorOut(errors.Wrap(err, "opening db"))
+	}
+	defer lggr.ErrorIfClosing(ldb, "db")
+
+	app, err := cli.AppFactory.NewApplication(cli.Config, ldb.DB())
```

Comment:
how does the application know what port number to use? It seems like the script is smart enough to start each node using correct port number (i.e. 6688, 6689, etc) from the demo but wasn't sure where the magic is happening.

Reply:
It's probably opening on 6688 every time, but we're not actually starting the nodes to be run. Just configuring them and then shutting them down. The dev can then set their environments and open up the configured nodes (or we can add another script :p)